astropy/coordinates/spectral_coordinate.py

```diff
@@ -0,0 +1,860 @@
+        # on the input frame to maintain rv/redshift continuity.
+        if self.target is None:
+            self._target = self._target_from_observer(
+                value, self.radial_velocity)
+
+        self._observer = value
+
+    @property
+    def target(self):
+        """
+        The coordinates of the target being observed.
+
+        Returns
+        -------
+        `~astropy.coordinates.BaseCoordinateFrame`
+            The astropy coordinate frame representing the target.
+        """
+        if self._target_specified:
+            return self._target
+
+    @target.setter
+    def target(self, value):
+        if self.target is not None:
+            raise ValueError("target has already been set")
+
+        self._target_specified = value is not None
+
+        value = self._validate_coordinate(value)
+
+        self._target = value
+
+    @property
+    def radial_velocity(self):
+        """
+        Radial velocity of target relative to the observer.
+
+        Returns
+        -------
+        `~astropy.units.Quantity`
+            Radial velocity of target.
+
+        Notes
+        -----
+        This is different from the ``.radial_velocity`` property of a
+        coordinate frame in that this calculates the radial velocity with
+        respect to the *observer*, not the origin of the frame.
+        """
+        return self._calculate_radial_velocity(self._observer, self._target,
+                                               as_scalar=True)
+
+    @property
+    def redshift(self):
+        """
+        Redshift of target relative to observer. Calculated from the radial
+        velocity.
+
+        Returns
+        -------
+        float
+            Redshift of target.
+        """
+        return _velocity_to_redshift(self.radial_velocity)
+
+    @staticmethod
+    def _calculate_radial_velocity(observer, target, as_scalar=False):
+        """
+        Compute the line-of-sight velocity from the observer to the target.
+
+        Parameters
+        ----------
+        observer : `~astropy.coordinates.BaseCoordinateFrame`
+            The frame of the observer.
+        target : `~astropy.coordinates.BaseCoordinateFrame`
+            The frame of the target.
+        as_scalar : bool
+            If `True`, the magnitude of the velocity vector will be returned,
+            otherwise the full vector will be returned.
+
+        Returns
+        -------
+        `~astropy.units.Quantity`
+            The radial velocity of the target with respect to the observer.
+        """
+
+        # Convert observer and target to ICRS to avoid finite differencing
+        # calculations that lack numerical precision.
+        observer_icrs = observer.transform_to(ICRS)
+        target_icrs = target.transform_to(ICRS)
+
+        pos_hat = SpectralCoord._norm_d_pos(observer_icrs, target_icrs)
+
+        d_vel = target_icrs.velocity - observer_icrs.velocity
```

Comment:
I'm getting a bit brain-dead as I review this quickly, but shouldn't this use the relativistic addition formula?

Reply:
I think so, and I think there are a few other places where this is needed too. Since I'm short on time in terms of doing this before feature freeze, I've opened an issue about it here:

https://github.com/astropy/astropy/issues/10249

I think it will be easier to focus on that separately from all the other implementation details here next week, and since that is just fixing numerical accuracy rather than a new feature, it can be merged after feature freeze or make it in a bugfix release if needed.

Reply:
@astrofrog - yes, this makes sense to put off to a follow-up bug-fix. I think it might be good to have a little helper function that can also deal with representations.